packages/captp/src/captp.js

```diff
@@ -443,30 +443,43 @@ export const makeCapTP = (
 
       const [method, args] = unserialize(serialized);
 
+      const DONE_PUMPKIN = { toString: () => 'DONE_PUMPKIN' };
       const getNextResultP = async () => {
         const result = await resultP;
-        if (!result || result.done) {
-          // We're done!
-          trapIterator.delete(questionID);
-          trapIteratorResultP.delete(questionID);
-          return result;
-        }
+        try {
+          if (!result || result.done) {
+            throw DONE_PUMPKIN;
+          }
+
+          const ait = trapIterator.get(questionID);
+          if (!ait) {
+            // The iterator is done, so we're done.
+            throw DONE_PUMPKIN;
+          }
 
-        const ait = trapIterator.get(questionID);
-        if (ait && ait[method]) {
           // Drive the next iteration.
-          return ait[method](...args);
+          return await ait[method](...args);
+        } catch (e) {
```

Comment:
I don't understand the new error logic. Why do we throw and catch the DONE_PUMPKIN in the same function for a non-exceptional case, rather than doing the non-exceptional case non-exceptionally?

Reply:
That was my convoluted way of reusing the cleanup logic.  I will just define a `cleanup()` function and remove the convolution.

Reply:
Much better! Resolving.